examples/postgres/custom_arrays/README.md

```diff
@@ -0,0 +1,892 @@
+all tests in the test folder.
+
+## Applied Best Practices
+
+Implementing the database operations on the Service type is simple, straight forward,
+and follows three best practices:
+
+1) Take the database connection as a parameter
+2) Prefer the generated DSL over custom queries whenever possible
+3) Add type annotations on database return types
+
+### Database connection as a parameter
+
+The first one refers to the idea that no type should hold application state, and a database connection definitely counts as application state. Instead, you would write a database connection manager 
+that manages a Postgres connection pool, and then calls into the database methods implemented 
+in the service type giving it a pooled connection as parameter. 
+
+
+### Prefer the generated DSL over custom queries
+
+The second one refers to the idea that non-complex queries are usually easier and more composable expressed in the
+generated DSL than in custom SQL queries and therefore the DSL is preferred in that case. However, if you have to write
+complex transactions, there is a chance the DSL isn’t cutting it, so in that scenario you may end up writing custom SQL.
+That said, the DSL gets you a long way before you are writing custom SQL.
+
+### Add type annotations
+
+The last one, type annotation on database return types, isn’t that obvious because Rust can infer the database return
+type from the method return type, so why bother? Why would you write get_result::<Service>(db) instead of just
+get_result(db) you may wonder? Here is the big secret, type inference in Rust needs occasionally a hint. When you
+annotated the database return type and then wants to daisy chain another operation via map or flatmap, without the type
+annotation you get a compile error. If you have the type annotation in place, you can add more operations whenever you
+want and things compile right away.
+To illustrate the last point, the count operator in Postgres only returns an i64 integer. Suppose your application
+expects an u64 from the persistence API. Wouldn’t it be nice to make the conversion on the fly?
+
+Let’s take a closer look. Conventionally, you write the count operation like so:
+
+```rust
+  pub fn count(db: &mut Connection) -> QueryResult<i64> {
+        service.count().get_result::<i64>(db)
+    }
+```
+
+Notice, the get result has a type annotation for i64. If you remove it, the code still compiles. But leave it there for
+a moment. Let’s add a map operation that takes an i64 returned from the count query and converts it into an u64. Notice,
+we have to update the return type of the method as well.
+
+```rust
+   pub fn count_u64(db: &mut Connection) -> QueryResult<u64> {
+        service.count()
+        .get_result::<i64>(db)
+        .map(|c| c as u64)
+    }
+```
+
+Run the compile and see that it works. Now, if you were to remove the type annotation <i64> from get_result, you get a
+compile error saying that trait FromSql is not implemented.
```

Comment:
Just to add a bit more context here: That's sometimes requires because there can be more than one `FromSql` impl for the same rust type and there can be more than one `FromSql` type for the same sql type. That gives a lot flexibility how you actually map data between your database and your rust code, but it requires that you be explicit about these types.